leakpro/synthetic_data_attacks/anonymeter/stats/confidence.py

```diff
@@ -0,0 +1,228 @@
+    confidence which is robust in problematic cases (e.g., when p goes extreme or
+    sample size is small). The estimated rate is a weighted average between the
+    MLE result and 0.5 which, however, in the sample sizes used in privacy attacks
+    does not differ visibly from the MLE outcome.
+
+    Parameters
+    ----------
+    n_total : int
+        Size of the sample.
+    n_success : int
+        Number of successful trials in the sample.
+    confidence_level : float
+        Confidence level for the error estimation.
+
+    Returns
+    -------
+    SuccessRate estimate (rate, error and confidence interval)
+
+    Notes
+    -----
+    E.B. WILSON
+    Probable inference, the law of succession, and statistical inference
+    Journal of the American Statistical Association 22, 209-212 (1927)
+    DOI 10.1080/01621459.1927.10502953
+
+    """
+    #Assert input
+    assert isinstance(n_total, int)
+    assert n_total > 0
+    assert isinstance(n_success, int)
+    assert n_success >= 0
+    assert isinstance(confidence_level, float)
+    if n_success > n_total:
+            raise ValueError("Parameter n_sucess can not be larger than n_total.")
+    assert isinstance(confidence_level, float)
+    assert_x_in_bound(x=confidence_level, x_name="confidence_level")
+    # Probit value for given confidence level
+    z = norm.ppf(0.5 * (1.0 + confidence_level))
+    # Calculations
+    z_squared = z * z
+    denominator = n_total + z_squared
+    rate = (n_success + z_squared / 2) / denominator
+    n_success_var = n_success * (n_total - n_success) / n_total
+    error = (z / denominator) * sqrt(n_success_var + z_squared / 4)
+    return SuccessRate(rate=rate, error=error)
+
+def residual_rate(*,
+    main_rate: SuccessRate,
+    naive_rate: SuccessRate,
+) -> SuccessRate:
+    """Compute residual success rate in a privacy attack.
+
+    Residual success is defined as the excess of main attack
+    success over naive attack success, normalized w.r.t.
+    the margin of improvement (unsuccessful guesses rate of naive attack).
+
+    Parameters
+    ----------
+    main_rate : SuccessRate
+        Success rate of main attack.
+    naive_rate : SuccessRate
+        Success rate of naive attack.
+
+    Returns
+    -------
+    SuccessRate
+        Residual success rate, adjusted in case naive_rate>attack_rate,
+        in which case the estimate is 0.
+        The error estimate is the propagated error bound of the residual
+        success rate.
+        Note: if naive_rate.rate == 1, residual success rate cannot be determined
+        and a SucessRate with rate=1.0 will be returned.q
+
+    """
+    # Check for naive_rate == 1
+    if naive_rate.rate == 1:
+        warnings.warn("Success of naive attack is 100%. Cannot measure residual success rate.", stacklevel=2)
+        return SuccessRate(rate=1.0, error=0.01)
+    # The calculations below (both expectation and error propagation) are incorrect
```

Comment:
I can only spot it being erroneous for one of the derivatives in the error propagation

Reply:
we are aware that calculations are incorrect, will continue to investigate but merge for moment being